webapp/dojo_news/src/App.spec.js

```diff
@@ -0,0 +1,221 @@
+import {shallowMount} from '@vue/test-utils';
+import App from "./App";
+import 'regenerator-runtime';
+import Vue from "vue";
+
+describe("News list", () => {
+    it('does not render any items on an initially empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+    });
+
+    it('renders items on an initially filled list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                    {id: 1, title: "Linux", votes: 0},
+                    {id: 2, title: "Windows", votes: 0}
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+
+        expect(wrapper.findAll(".newslistitem").length).toBe(3);
+    });
+    it('renders items correctly dependent on mutations of the list of NewsItems', async () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(1);
+
+        wrapper.vm.removeNewsListItem(0);
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+
+        wrapper.vm.createNewsListItem("Test");
+        wrapper.vm.createNewsListItem("Test2");
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(2);
+    });
+});
+
+describe("List Empty Message", () => {
+    const listEmptyMessage = "Oh noes, the list is empty!";
+
+    it('displays a message on an initially empty list of NewsItems', () => {
+        const wrapperWithEmptyList = shallowMount(App, {
+            propsData: {
+                listEmptyMessage: listEmptyMessage,
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapperWithEmptyList.text()).toContain(listEmptyMessage);
```

Comment:
```suggestion
        expect(wrapperWithEmptyList.text()).toContain("Oh noes, the list is empty!");
```

I prefer hard-coded assertion values.

Reply:
Is that a request for changes or an annotation? :D And wouldn't this also require all the tests to be adjusted, when the list empty message gets changed?

Reply:
Yes, that' correct, you have to touch your tests when you change the empty message.

The reason is a trade-off: I ran into false-negatives in the past when I used assertion values based on a variable. An example is e..g empty string is always contained in any text. If, for some reason, you made a mistake in your test setup you may run into a silently passing (but false) test cases. Also, you lower the complexity of your test case, making it easier to read.

Failing tests because of a changed empty message would be a false-positive in this definition. Yet, false-negatives are a lot *worse* than false-positives, that's why I suggest to use hard-coded assertion values where possible.

Reply:
"An example is e..g empty string is always contained in any text." - eye opening, thanks! Fixed it. However you probably mean false-positives when you write false-negatives and vice versa. If you were to check for an empty string by mistake and the test would pass because of that, the result is positive, but that is false. So you have a false-positive, which is indeed a lot worse than a false-negative.

Reply:
Depends on your definition: I use the analogy of a [Corona|HIV]-Test: It's negative when the test says, you are not infected. So a software test is negative, when it's passing. When you have a test case passing when it should be failing, this would be a false-negative according to this definition.

Reply:
Right.